File: Baekjoon/jeongyoung/나무 자르기.cpp
#include <iostream>
#include <vector>

#define ll long long 
using namespace std;

int N, M;
vector<int> arr;

int main(){

  cin >> N >> M;
  for(int i = 0 ; i < N; i++){
    int a;
    cin >> a;
    arr.push_back(a);
  }

  int low = 0;
  int high = 1e9;
  int ans = 1e9;

  int mid = (low + high) / 2;
  while(low < high) {
    mid = (low + high) / 2;

    ll sum = 0;
    for(int i = 0 ; i < N; i++){
      if(arr[i] - mid > 0 )
        sum += arr[i] - mid;
    }
    //cout << high << " " << low << " " << mid << " " << sum << "\n";

    if(sum >= M){
      ans = mid;
      low = mid + 1;
    }else{
      high = mid;
    }
  }

  cout << ans;


  return 0;
}

File: Baekjoon/eunnyeong/나무자르기.java
import java.io.*;
import java.util.*;

public class Main {

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    StringTokenizer st = new StringTokenizer(br.readLine());
    int n = Integer.parseInt(st.nextToken());
    int m = Integer.parseInt(st.nextToken());

    int[] a = new int[n];
    st = new StringTokenizer(br.readLine());
    for (int i = 0; i < n; i++)
      a[i] = Integer.parseInt(st.nextToken());

   Arrays.sort(a);

   int low = 0, high = a[n - 1];
   while (low < high) {
     int mid = (low + high) / 2;
     long sum = 0;

     for (int i = n - 1; i >= 0; i--) {
       if (a[i] > mid)
         sum += a[i] - mid;
       else break;
     }

     if (sum < m)
       high = mid;
     else
       low = mid + 1;
   }

   System.out.print(low - 1);
  }
}
